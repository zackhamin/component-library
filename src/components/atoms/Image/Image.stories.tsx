import { Image, ImageProps } from './Image';
import { Meta, Story } from '@storybook/react';
import TestSvg from 'Atoms/Image/assets/TestSvg';

export default {
	title: 'atoms/Image',
	component: Image,
} as Meta;

const defaultArgs: ImageProps<'img'> = {
	as: 'img',
	src: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAnklEQVR42u3RAQ0AAAQAMLro30oOapj9FZ5dNcEZKUQIQoQgRAhChCBEiBAhCBGCECEIEYIQIQhBiBCECEGIEIQIQQhChCBECEKEIEQIQhAiBCFCECIEIUIQghAhCBGCECEIEYIQhAhBiBCECEGIEIQgRAhChCBECEKEIAQhQhAiBCFCECIEIQgRghAhCBGCECEIESJECEKEIEQIQr5bkNXZleU60yAAAAAASUVORK5CYII=',
	alt: 'An image',
	title: 'An image',
	loading: 'lazy',
	block: false,
};

const Template: Story = (args) => <Image {...args} />;

export const Default = Template.bind({});
Default.args = defaultArgs;

export const BlockImage = Template.bind({});
BlockImage.args = { ...defaultArgs, block: true };

export const SvgImage = Template.bind({});
SvgImage.args = {
	as: 'svg',
	width: '500',
	height: '135',
	viewBox: '0 0 1229 135',
	fill: 'none',
	xmlns: 'http://www.w3.org/2000/svg',
	style: {
		padding: '1.4rem',
		background: '#00594c',
	},
	block: false,
	children: TestSvg,
} as ImageProps<'svg'>;

export const ObjectImage = Template.bind({});
ObjectImage.args = {
	as: 'object',
	data: 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAnklEQVR42u3RAQ0AAAQAMLro30oOapj9FZ5dNcEZKUQIQoQgRAhChCBEiBAhCBGCECEIEYIQIQhBiBCECEGIEIQIQQhChCBECEKEIEQIQhAiBCFCECIEIUIQghAhCBGCECEIEYIQhAhBiBCECEGIEIQgRAhChCBECEKEIAQhQhAiBCFCECIEIQgRghAhCBGCECEIESJECEKEIEQIQr5bkNXZleU60yAAAAAASUVORK5CYII=',
	width: 640,
	height: 360,
	margin: 'auto',
	block: false,
} as ImageProps<'object'>;
