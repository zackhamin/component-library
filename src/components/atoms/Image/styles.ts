import styled from 'styled-components';

const StyledImage = styled.img`
	&.block {
		width: 100%;
		display: block;
	}
`;

export { StyledImage };
