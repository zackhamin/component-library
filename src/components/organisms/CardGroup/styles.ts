import styled from 'styled-components';

const StyledCardGroup = styled.div`
	display: grid;
	max-width: 100%;
	grid-template-columns: repeat(auto-fit, minmax(314px, 1fr));
	gap: 1.6rem;
`;

export { StyledCardGroup };
