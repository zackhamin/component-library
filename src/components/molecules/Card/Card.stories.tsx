import { Button } from 'Atoms/Button';
import { Heading } from 'Atoms/Heading';
import { Image } from 'Atoms/Image';
import { Link } from 'Atoms/Link';
import { Select } from 'Atoms/Select';
import TestSvg from 'Atoms/Image/assets/TestSvg';
import { Text } from 'Atoms/Text';
import { Card, CardProps } from './Card';
import { Meta, Story } from '@storybook/react';

export default {
	title: 'molecules/Card',
	component: Card,
} as Meta;

const Template: Story<CardProps> = (args) => {
	return <Card {...args}>{args.children}</Card>;
};

const defaultArgs = {
	image: (
		<Image
			as="img"
			src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAnklEQVR42u3RAQ0AAAQAMLro30oOapj9FZ5dNcEZKUQIQoQgRAhChCBEiBAhCBGCECEIEYIQIQhBiBCECEGIEIQIQQhChCBECEKEIEQIQhAiBCFCECIEIUIQghAhCBGCECEIEYIQhAhBiBCECEGIEIQgRAhChCBECEKEIAQhQhAiBCFCECIEIQgRghAhCBGCECEIESJECEKEIEQIQr5bkNXZleU60yAAAAAASUVORK5CYII="
			alt="An image"
			title="An image"
			loading="lazy"
		/>
	),
};

export const Default = Template.bind({});
Default.args = {
	image: defaultArgs.image,
	children: (
		<>
			<Heading size={'2.2rem'} level={3}>
				This has
			</Heading>
			<Text>An Image</Text>
		</>
	),
};

export const NoImage = Template.bind({});
NoImage.args = {
	children: (
		<>
			<Heading size={'2.2rem'} level={3}>
				Whilst this has
			</Heading>
			<Text>No Image</Text>
		</>
	),
};

export const ImageWithLink = Template.bind({});
ImageWithLink.args = {
	image: defaultArgs.image,
	children: (
		<Link
			as={'a'}
			href={'https://www.hollandandbarrett.com/'}
			target={'_blank'}
		>
			<Heading size={'2.2rem'} level={3}>
				Image with link
			</Heading>
			<Text>Please click the link</Text>
		</Link>
	),
};

export const InputComponents = Template.bind({});
InputComponents.args = {
	image: defaultArgs.image,
	children: (
		<>
			<Select options={[1, 2, 3, 4, 5]} />
			<br />
			<Button title={'Click here'}>Click here</Button>
		</>
	),
};

export const CustomImage = Template.bind({});
CustomImage.args = {
	image: (
		<Image
			as="svg"
			width="1229"
			height="135"
			viewBox="0 0 1229 135"
			fill="none"
			xmlns="http://www.w3.org/2000/svg"
			style={{
				padding: '1.4rem',
				background: '#00594c',
			}}
		>
			{TestSvg}
		</Image>
	),
	children: (
		<Link href={'https://www.hollandandbarrett.com/'} target={'_blank'} block>
			<Button block>Visit HnB</Button>
		</Link>
	),
};
