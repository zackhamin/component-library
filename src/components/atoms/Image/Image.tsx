import classNames from 'classnames';
import { StyledImage } from './styles';
import { ElementType, forwardRef } from 'react';
import {
	PolymorphicComponentProps,
	PolymorphicRef,
} from '@Types/polymorphism-utils';

export type ImageProps<Component extends ElementType> =
	PolymorphicComponentProps<Component> & {
		block?: boolean;
	};

const Image = forwardRef(
	<Component extends ElementType = 'img'>(
		{ as, block, ...props }: ImageProps<Component>,
		ref?: PolymorphicRef<Component>
	) => {
		const classes = classNames({
			block,
		});

		const tag = as || 'img';
		return <StyledImage ref={ref} as={tag} {...props} className={classes} />;
	}
);

export { Image };
