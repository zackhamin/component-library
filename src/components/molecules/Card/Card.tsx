import { StyledCard } from './styles';
import { ReactElement, ReactNode } from 'react';

export type CardProps = {
	image?: ReactElement<HTMLOrSVGImageElement | HTMLObjectElement | unknown>;
	children: ReactNode;
};

const Card = ({ children, image }: CardProps) => {
	if (image) {
		return (
			<StyledCard>
				<div className={'card-image'}>{image}</div>
				<figcaption className={'card-text'}>{children}</figcaption>
			</StyledCard>
		);
	}

	return (
		<StyledCard as={'div'}>
			<div className={'card-text'}>{children}</div>
		</StyledCard>
	);
};

export { Card };
