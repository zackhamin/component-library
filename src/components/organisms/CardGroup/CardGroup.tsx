import { StyledCardGroup } from './styles';
import { ReactElement } from 'react';

export type CardGroupProps = {
	children: ReactElement | ReactElement[];
};

const CardGroup = ({ children }: CardGroupProps) => {
	return <StyledCardGroup>{children}</StyledCardGroup>;
};

export { CardGroup };
