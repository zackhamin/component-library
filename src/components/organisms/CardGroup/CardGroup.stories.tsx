import { CardGroup, CardGroupProps } from './CardGroup';
import { Card } from 'Molecules/Card';
import { Text } from 'Atoms/Text';
import { Image } from 'Atoms/Image';

export default {
	title: 'organisms/CardGroup',
	component: CardGroup,
};

export const Template = (args: CardGroupProps) => <CardGroup {...args} />;

Template.storyName = 'CardGroup';
Template.args = {
	children: (
		<>
			<Card
				image={
					<Image
						src={
							'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAnklEQVR42u3RAQ0AAAQAMLro30oOapj9FZ5dNcEZKUQIQoQgRAhChCBEiBAhCBGCECEIEYIQIQhBiBCECEGIEIQIQQhChCBECEKEIEQIQhAiBCFCECIEIUIQghAhCBGCECEIEYIQhAhBiBCECEGIEIQgRAhChCBECEKEIAQhQhAiBCFCECIEIQgRghAhCBGCECEIESJECEKEIEQIQr5bkNXZleU60yAAAAAASUVORK5CYII='
						}
					/>
				}
			>
				<Text>Red</Text>
			</Card>
			<Card
				image={
					<Image
						src={
							'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAoElEQVR42u3RAQ0AMAgAoNvlwSxvDq3hHFQgflY/1gghQhAiBCFCECIEIUKECEGIEIQIQYgQhAhBCEKEIEQIQoQgRAhCECIEIUIQIgQhQhCCECEIEYIQIQgRghCECEGIEIQIQYgQhCBECEKEIEQIQoQgBCFCECIEIUIQIgQhCBGCECEIEYIQIQhBiBCECEGIEIQIQYgQIUIQIgQhQhBy3QA05/EFbfI86QAAAABJRU5ErkJggg=='
						}
					/>
				}
			>
				<Text>Blue</Text>
			</Card>
			<Card
				image={
					<Image
						src={
							'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAoUlEQVR42u3RAQ0AMAgAoNvlFeyfxhxawzmoQPzKfqwRQoQgRAhChCBECEKECBGCECEIEYIQIQgRghCECEGIEIQIQYgQhCBECEKEIEQIQoQgBCFCECIEIUIQIgQhCBGCECEIEYIQIQhBiBCECEGIEIQIQQhChCBECEKEIEQIQhAiBCFCECIEIUIQghAhCBGCECEIEYIQIUKEIEQIQoQg5LoBnELgOTqoyNsAAAAASUVORK5CYII='
						}
					/>
				}
			>
				<Text>Green</Text>
			</Card>
			<Card
				image={
					<Image
						src={
							'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAnUlEQVR42u3RAQ0AAAgDIO3fzUq3hnNQgc5UijNaiBCECEGIEIQIQYgQIUIQIgQhQhAiBCFCEIIQIQgRghAhCBGCEIQIQYgQhAhBiBCEIEQIQoQgRAhChCAEIUIQIgQhQhAiBCEIEYIQIQgRghAhCEGIEIQIQYgQhAhBCEKEIEQIQoQgRAhCECIEIUIQIgQhQhAiRIgQhAhBiBCEfLdGDBgklKcbOgAAAABJRU5ErkJggg=='
						}
					/>
				}
			>
				<Text>Yellow</Text>
			</Card>
		</>
	),
};
