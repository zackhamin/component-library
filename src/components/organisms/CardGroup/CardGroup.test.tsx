import { CardGroup } from '../CardGroup';
import { render } from '@testing-library/react';
import { Card } from 'Molecules/Card';
import { Image } from 'Atoms/Image';

describe('CardGroup', () => {
	it('renders a card group correctly', () => {
		const { container } = render(
			<CardGroup>
				<Card image={<Image as={'img'} />}>Child</Card>
				<Card>child</Card>
				<Card>child</Card>
				<Card image={<Image as={'img'} />}>Child</Card>
			</CardGroup>
		);
		expect(container).toMatchSnapshot();
	});
});
