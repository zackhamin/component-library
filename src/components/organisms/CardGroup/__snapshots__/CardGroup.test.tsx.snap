// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardGroup renders a card group correctly 1`] = `
.c0 {
  display: grid;
  max-width: 100%;
  grid-template-columns: repeat(auto-fit,minmax(314px,1fr));
  gap: 1.6rem;
}

.c1 {
  border-radius: 0.8rem;
  overflow: hidden;
  padding: 0;
  margin: 0;
  text-align: left;
  box-shadow: 0 0.2rem 0.6rem #33333324;
  position: relative;
  -webkit-transition: box-shadow 0.2s;
  transition: box-shadow 0.2s;
}

.c1:hover {
  box-shadow: 0 0.8rem 1.6rem #33333324;
}

.c1 .card-text {
  padding: 1.4rem;
  margin: 0;
  width: inherit;
}

.c1 .card-text a {
  text-align: inherit !important;
  width: inherit;
}

.c1 .card-text a:before {
  position: absolute;
  width: 100%;
  height: 100%;
  left: 0;
  top: 0;
  content: '';
}

.c1 .card-image img {
  aspect-ratio: 314/154;
  object-fit: cover;
  object-position: center;
  position: relative;
  display: block;
  width: 100%;
}

.c2.block {
  width: 100%;
  display: block;
}

<div>
  <div
    class="c0"
  >
    <figure
      class="c1"
    >
      <div
        class="card-image"
      >
        <img
          class="c2"
        />
      </div>
      <figcaption
        class="card-text"
      >
        Child
      </figcaption>
    </figure>
    <div
      class="c1"
    >
      <div
        class="card-text"
      >
        child
      </div>
    </div>
    <div
      class="c1"
    >
      <div
        class="card-text"
      >
        child
      </div>
    </div>
    <figure
      class="c1"
    >
      <div
        class="card-image"
      >
        <img
          class="c2"
        />
      </div>
      <figcaption
        class="card-text"
      >
        Child
      </figcaption>
    </figure>
  </div>
</div>
`;
